controllers/machinedeletionremediation_controller.go

```diff
@@ -76,73 +99,123 @@ type MachineDeletionRemediationReconciler struct {
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
-func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (finalResult ctrl.Result, finalErr error) {
 	log := r.Log.WithValues("machinedeletionremediation", req.NamespacedName)
 
 	log.Info("reconciling...")
 
 	var err error
 	var remediation *v1alpha1.MachineDeletionRemediation
 	if remediation, err = r.getRemediation(ctx, req); remediation == nil || err != nil {
-		if apiErrors.IsNotFound(err) {
-			return ctrl.Result{}, nil
-		}
 		return ctrl.Result{}, err
 	}
 
-	annotations := remediation.GetAnnotations()
-	if annotations != nil {
-		machineNameNamespace, exists := annotations[MachineNameNamespaceAnnotation]
-		if exists {
-			machineName, machineNamespace, err := extractNameAndNamespace(machineNameNamespace, remediation.GetName())
-			if err != nil {
-				log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", machineNameNamespace)
-				return ctrl.Result{}, nil
-			}
-
-			ok, err := r.verifyMachineIsDeleted(ctx, machineName, machineNamespace, remediation.Name)
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-			if !ok {
-				return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	defer func() {
+		if updateErr := r.updateStatus(ctx, remediation); updateErr != nil {
+			if !apiErrors.IsConflict(updateErr) {
+				finalErr = utilerrors.NewAggregate([]error{updateErr, finalErr})
+			} else {
+				// requeue if there is no error already set
+				if finalErr == nil {
+					// if requeue is already set choose the lowest delay
+					minRequeue := time.Second
+					if finalResult.RequeueAfter > 0 && minRequeue > finalResult.RequeueAfter {
+						minRequeue = finalResult.RequeueAfter
+					}
+					finalResult.RequeueAfter = minRequeue
+				}
 			}
-			return ctrl.Result{}, nil
 		}
-	}
+	}()
 
-	var node *v1.Node
-	if node, err = r.getNodeFromMdr(remediation); err != nil {
-		return ctrl.Result{}, err
+	if r.isStoppedByNHC(remediation) {
+		log.Info("MDR remediation was stopped by Node Healthcheck")
+		return ctrl.Result{}, r.updateConditions(remediationTerminatedByNHC, remediation)
 	}
 
-	log.Info("MDR CR and affected Node found", "node", node.GetName())
+	state := r.getCurrentState(remediation)
```

Comment:
Looking at the CR's status for deciding what to do is an anti pattern and should be avoided. You should always look at the state in the cluster. This "state machine" based on multiple conditions makes the flow more complicated than needed IMHO. E.g., what's the value of just updating conditions for `remediationStarted` while nothing happened actually yet in that reconcile run?
I think I would keep the old flow and just set conditions when needed. E.g., at the beginning set Processing = true. When you see that the machine was deleted, set Processing = false, and Succeeded = true. In case you run into an unrecoverable error (so no requeue anymore) set Processing = false and Succeeded = false. That's all we need IMHO. WDYT?

Reply:
Ok, I understand the problem. And yes, in this case it sounds ok to me to just update the processing conditon as a first step. 

> I'll revert it to the simpler implementation, trying to make it clearer.

sounds like a plan, thanks! 👍🏼

Reply:
or maybe just like this, updating annotation and condition one after another? 🤔
 
- if annotation doesn't exist, set annotation and update (without requeue in case it succeeds)
- if annotation exists (on the potentially updated CR), and processing condition is not set yet, set condition and requeue

Reply:
in case the condition update fails, we have a not up to date CR... but that will be fixed in the next requeue, so no issue IMHO, also because the actual remediation didn't start yet.

Reply:
This will work, the drawback is that the deferred function that handles the status.update is kind of useless in this case, while it's used in the other updates 🤷

I mean, using a function to update status.condition immediately, instead than relying only on the automatic call to the deferred function, works very well. I can just update first the status.condition (also in the cluster) even before that the annotation is set locally

Reply:
not sure if your 1st sentence still is valid, it's not striked through 😎 

>  the drawback is that the deferred function that handles the status.update is kind of useless in this case, while it's used in the other updates

I don't see why it's useless?

Reply:
I should've striked more :smile: 
- kind of useless: update the `status.condition` before the annotation is to be done _in the cluster_ too, so I couldn't wait the deferred function to do it, for which *in this case* was not useful

But this is an old problem, I see now how you're idea work (cool that it was also one of the first implementations :facepalm:)

- request machine deletion
- update annotation, both locally and in the cluster, get the new mdr
- set status.condition locally
- let the deferred function update status.condition in the cluster

I think this is the way :tm:

Reply:
almost, I would request machine deletion AFTER annotation and condition were set...
simplified pseudo code without error handling (just return every error immediately):
```
if ! annotation exists {
  set annotation
  update CR
}
if ! condition set {
  set condition (processing true)
  requeue (deferred status update)
}
if machine exists {
  if !deletionTimestamp set {   
    delete machine
    requeue
  } 
  // maybe check for some kind of timeout? -> set processing false, succeeded false, return without requeue
  // otherwise just requeu
}
if !condition set
  set condition (processing false, succeeded true)
}
done (return without requeue (deferred status update))
```

Reply:
> almost, I would request machine deletion AFTER annotation and condition were set...

could I ask you what the rationale is?

Consider that the annotation has the only purpose to check if the machine exists after deletion request, without relying to the existence of its node. For this reason, once is set, we check if the machine exists and if so, requeue (no machine deletion request).

Is it Ok to request machine deletion multiple times?

Reply:
>> almost, I would request machine deletion AFTER annotation and condition were set...

> could I ask you what the reason is?

you want to save the machine name before deleting the machine, not? 🤔 

> Is it Ok to request machine deletion multiple times?

I don't see that in my suggested flow. Machine is deleted only when

- the annotation exists
- the condition was set
- the machine exists
- and it doesn't have a deletion timestamp

Reply:
> I don't see that in my suggested flow. Machine is deleted only when

I read wrong the `deletionTimestamp` part. I thought it was MDR's `deletionTimestamp

> you want to save the machine name before deleting the machine, not? thinking

Correct, but the annotation is used to track the machine's status during it's deletion [here](https://github.com/medik8s/machine-deletion-remediation/blob/main/controllers/machinedeletionremediation_controller.go#L93), so once the annotation is set, and we requeue, MDR expects deletion was requested already

Reply:
>  once the annotation is set, and we requeue, MDR expects deletion was requested already

and that's not an ideal approach. Look at the machine if it's deleted, not the annotation. That's why my  suggested flow is different from what we currently have in the code

Reply:
I think I'll rework this in a different branch, then I'll rebase this PR

Reply:
or just open a new PR 😉

Reply:
Working on it 😁